File: sliding-window/1456-Maximum-Number-of-Vowels-in-a-Substring-of-Given-Length.go
func maxVowels(s string, k int) int {
    left := 0
    res := 0
    current_vowel := 0

    for right, _ := range s {
        if isVowel(rune(s[right])) {
            if isVowel(rune(s[right])) {
                current_vowel++
            }
        }

        for right - left + 1 > k {
            if isVowel(rune(s[left])) {
                current_vowel--
            }
            left++
        }

        res = max(res, current_vowel)
    }
    return res
}

func isVowel(c rune) bool {
    if c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' {
        return true
    }
    return false
}

func max(a int, b int) int {
    if a > b {
        return a
    }
    return b
}

File: reference/sliding-window/1456-Maximum-Number-of-Vowels-in-a-Substring-of-Given-Length.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowel_letters = {"a", "e", "i", "o", "u"}  

        current_vowel = 0
        max_number, left = 0, 0
        for right in range(len(s)):
            if s[right] in vowel_letters:
                current_vowel += 1
            if right - left + 1 > k:
                if s[left] in vowel_letters:
                    current_vowel -= 1
                
                left += 1
            
            max_number = max(max_number, current_vowel)
        return max_number

